feat: add command that can restore a cluster from a backup id

oleschoenburg opened the pull request:
This PR adds `backup setup` and `backup restore` commands and fix a couple of issues found while testing.

oleschoenburg commented:
@Zelldon Apologies for the large PR. I'd recommend to go commit by commit but skip [c4058c3](https://github.com/zeebe-io/zeebe-chaos/pull/213/commits/c4058c3e81ea67eaebbddfc2c540e489d73f533d) since this is just a grab-bag of refactorings and fixes of things found during testing.

Zelldon commented:
@oleschoenburg sorry for the delay it is on my plate for today/tomorrow

Zelldon commented on go-chaos/Dockerfile:
```diff
@@ -1,6 +1,6 @@
 # syntax=docker/dockerfile:1.4
 FROM golang:alpine as builder
-
+RUN apk update && apk upgrade && apk add --no-cache ca-certificates && update-ca-certificates
```
:+1:

Zelldon commented on go-chaos/Dockerfile:
```diff
@@ -9,12 +9,13 @@ RUN go mod download
 
 WORKDIR /app
 
+COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
```
TIL

Zelldon commented on go-chaos/cmd/backup.go:
```diff
@@ -58,12 +71,121 @@ var waitForBackupCommand = &cobra.Command{
+
+	namespace := k8Client.GetCurrentNamespace()
+
+	err = k8Client.PauseReconciliation()
```
:question: : What happens on our clusters here?

oleschoenburg replied:
Set's the `pauseReconciliation` flag to true which prevents the SaaS operator from reverting our changes to the stateful sets.

Zelldon replied:
But I have seen later you ignore it. But was not sure when reading first.

Zelldon commented on go-chaos/cmd/backup.go:
```diff
@@ -58,12 +71,121 @@ var waitForBackupCommand = &cobra.Command{
+		return err
+	}
+
+	_, err = copyBackupSecret(cmd, k8Client, namespace)
```
:question: for what is this?

Zelldon commented on go-chaos/cmd/backup.go:
```diff
@@ -58,12 +71,121 @@ var waitForBackupCommand = &cobra.Command{
+	sfsEnv := sfs.Spec.Template.Spec.Containers[0].Env
+	sfs.Spec.Template.Spec.Containers[0].Env = append(
+		sfsEnv,
+		core.EnvVar{Name: "ZEEBE_BROKER_DATA_BACKUP_STORE", Value: "S3"},
+		core.EnvVar{Name: "ZEEBE_BROKER_EXPERIMENTAL_FEATURES_ENABLEBACKUP", Value: "true"},
+		core.EnvVar{Name: "MANAGEMENT_ENDPOINTS_WEB_EXPOSURE_INCLUDE", Value: "*"},
+		core.EnvVar{Name: "MANAGEMENT_ENDPOINTS_BACKUPS_ENABLED", Value: "true"},
+	)
```
:bulb:

Zelldon commented on go-chaos/cmd/backup.go:
```diff
@@ -58,12 +71,121 @@ var waitForBackupCommand = &cobra.Command{
+}
+
+func setupGatewayForBackups(cmd *cobra.Command, err error, k8Client internal.K8Client, namespace string) error {
+	saasGatewayLabels := meta.LabelSelector{
+		MatchLabels: map[string]string{"app.kubernetes.io/component": "standalone-gateway"},
+	}
+	var gatewayDeployments *apps.DeploymentList
```
:x: this works only for SaaS would be great if we can get it running for all.

There is a method to get the gateway if I remember correctly, no matter which env.

Zelldon commented on go-chaos/cmd/backup.go:
```diff
@@ -58,12 +71,121 @@ var waitForBackupCommand = &cobra.Command{
+	if err != nil {
+		return err
+	}
+	if len(gatewayDeployments.Items) == 0 {
```
I saw this later, sorry. But still I think there is a getGateway function :thinking:

Zelldon commented on go-chaos/cmd/backup.go:
```diff
@@ -58,12 +71,121 @@ var waitForBackupCommand = &cobra.Command{
+	return err
+}
+
+func copyBackupSecret(cmd *cobra.Command, k8Client internal.K8Client, namespace string) (*core.Secret, error) {
```
:x: Maybe we can call it create, it was a bit confusing that it is called copy. I expected it a bit different.

Zelldon commented on go-chaos/cmd/worker.go:
```diff
@@ -71,22 +72,25 @@ func handleZbChaosJob(client worker.JobClient, job entities.Job) {
 		},
 	}
 	err := job.GetVariablesAs(&jobVariables)
+	//fmt.Printf("Handling %T \n", jobVariables)
```
:wrench: Can we remove this?

Zelldon commented on go-chaos/cmd/worker.go:
```diff
@@ -96,6 +100,7 @@ func handleZbChaosJob(client worker.JobClient, job entities.Job) {
 }
 
 func runZbChaosCommand(args []string, ctx context.Context) error {
+	fmt.Printf("Running command with args: %v \n", args)
```
:+1:

Zelldon commented on go-chaos/internal/k8helper.go:
```diff
@@ -29,12 +30,13 @@ import (
-	ClientConfig clientcmd.ClientConfig
-	Clientset    kubernetes.Interface
+	ClientConfig  clientcmd.ClientConfig
+	DynamicClient dynamic.Interface
```
:question: For what is this?

oleschoenburg replied:
we need a dynamic client to work with the SaaS CRDs. The normal k8s client we used for everything else only supports well known resources and can't be used for CRDs that it doesn't know about.

Zelldon replied:
Cool thanks :+1:

Zelldon commented on go-chaos/internal/cluster.go:
```diff
@@ -0,0 +1,120 @@
+	namespace := c.GetCurrentNamespace()
+	clusterId := strings.TrimSuffix(namespace, "-zeebe")
+	zeebeCrd := schema.GroupVersionResource{Group: "cloud.camunda.io", Version: "v1alpha1", Resource: "zeebeclusters"}
+	payload := fmt.Sprintf(`{"metadata": {"labels": {"cloud.camunda.io/pauseReconciliation": "%t"}}}`, pauseEnabled)
```
TIL :bulb: 

Interesting :+1: previously we had to update the CRD

oleschoenburg replied:
That's what this is. It uses the dynamic client to get the custom resource for the zeebe cluster and patches it with a new label.

Zelldon replied:
This label is new, before it was part of the crd

Zelldon commented on go-chaos/internal/cluster.go:
```diff
@@ -0,0 +1,120 @@
+	if err != nil {
+		return 0, err
+	}
+	err = wait.PollImmediateUntilWithContext(
```
:+1:

Zelldon commented on go-chaos/internal/cluster.go:
```diff
@@ -0,0 +1,120 @@
+		return 0, err
+	}
+	statefulSets := c.Clientset.AppsV1().StatefulSets(namespace)
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
:+1: could we use on more places I guess

oleschoenburg replied:
Yep :+1: It only added it here because some of the setup methods change resources in rapid succession where conflicts are much more likely to occur. But in general it's a good practice I think.

Zelldon commented on go-chaos/internal/cluster.go:
```diff
@@ -0,0 +1,120 @@
+	namespace := c.GetCurrentNamespace()
+	ctx := context.TODO()
+
+	helmLabel := meta.LabelSelector{
+		MatchLabels: map[string]string{"app.kubernetes.io/name": "zeebe"},
+	}
```
Please check I think we have a function to resolve the labels

Zelldon reviewed (changes requested):
Thanks @oleschoenburg :+1: Cool stuff :100: 

I have some comments and remarks please have a look :)

oleschoenburg commented:
Sorry, I didn't realize that you wanted to do a second review round so I already rebased to resolve some merge conflicts. I think I've fixed everything you pointed out, please take another look :pray:

Zelldon commented on go-chaos/cmd/backup.go:
```diff
@@ -58,12 +71,121 @@ var waitForBackupCommand = &cobra.Command{
+			return err
+		}
+	}
+	gateway := gatewayDeployments.Items[0]
```
Until here we have the same code already somewhere else (**checkIfGatewaysAreRunning**). But I can also clean that up later.

Zelldon reviewed (approved):
Thanks @oleschoenburg :bow:
